File: samples/java/SupportChat/Server/Api/src/main/java/systems/zlink/samples/supportchat/server/api/handlers/AuthenticateUserHandler.java
package systems.zlink.samples.supportchat.server.api.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.supportchat.server.configuration.SampleNames;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup(SampleNames.ApiChannel)
public final class AuthenticateUserHandler
        implements ZLinkRequestHandler<Messages.AuthenticateUserReq, Messages.AuthenticateUserRes> {
    @Override
    public CompletionStage<Messages.AuthenticateUserRes> handle(
            Messages.AuthenticateUserReq request, ZLinkMessageContext context) {
        return CompletableFuture.completedFuture(
                switch (request.accessToken()) {
                    case "customer-1" ->
                            accepted("customer-1", "Customer 1", SampleNames.Roles.Customer);
                    case "customer-2" ->
                            accepted("customer-2", "Customer 2", SampleNames.Roles.Customer);
                    case "customer-3" ->
                            accepted("customer-3", "Customer 3", SampleNames.Roles.Customer);
                    case "agent-1" -> accepted("agent-1", "Agent 1", SampleNames.Roles.Agent);
                    case "agent-2" -> accepted("agent-2", "Agent 2", SampleNames.Roles.Agent);
                    default ->
                            new Messages.AuthenticateUserRes(
                                    false, null, null, null, "invalid token");
                });
    }

    private static Messages.AuthenticateUserRes accepted(
            String actorId, String displayName, String role) {
        return new Messages.AuthenticateUserRes(true, actorId, displayName, role, null);
    }
}

File: samples/kotlin/SupportChat/Server/Api/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/api/handlers/AuthenticateUserHandler.kt
package systems.zlink.samples.kotlin.supportchat.server.api.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.samples.kotlin.supportchat.server.configuration.SampleNames
import systems.zlink.samples.kotlin.supportchat.server.configuration.SupportChatRoles
import systems.zlink.samples.kotlin.supportchat.shared.contracts.AuthenticateUserReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.AuthenticateUserRes

@ZLinkHandlerGroup(SampleNames.ApiChannel)
class AuthenticateUserHandler :
    ZLinkSuspendingRequestHandler<AuthenticateUserReq, AuthenticateUserRes> {
    override suspend fun handle(
        request: AuthenticateUserReq,
        context: ZLinkMessageContext,
    ): AuthenticateUserRes =
        when (request.accessToken) {
            "customer-1" -> accepted("customer-1", "Customer 1", SupportChatRoles.Customer)
            "customer-2" -> accepted("customer-2", "Customer 2", SupportChatRoles.Customer)
            "customer-3" -> accepted("customer-3", "Customer 3", SupportChatRoles.Customer)
            "agent-1" -> accepted("agent-1", "Agent 1", SupportChatRoles.Agent)
            "agent-2" -> accepted("agent-2", "Agent 2", SupportChatRoles.Agent)
            else -> AuthenticateUserRes(false, null, null, null, "Unknown support chat token.")
        }

    private fun accepted(actorId: String, displayName: String, role: String): AuthenticateUserRes =
        AuthenticateUserRes(true, actorId, displayName, role, null)
}
